docs/proposals/connector-protocol-v2.md

```diff
@@ -224,58 +226,53 @@ notify(StreamId, StreamName, PointOfRef) ->
         point_of_reference(PointOfRef)
     ]).
 
--spec message(Message, MessageId, StreamId, Flags) -> frame() when
-    Message :: boundary | {non_neg_integer(), iodata()} | iodata(),
-    MessageId :: non_neg_integer(),
+-spec message(Flags, StreamId, MessageId, EventTime, Key, Message) -> frame()
+  when
+    Flags :: non_neg_integer(),      % See message bit flag definitions above
     StreamId :: non_neg_integer(),
-    Flags :: non_neg_integer(). % See message bit flag definitions above
-message(_Message=boundary, MessageId, StreamId, Flags) ->
-    frame([
-        ?MESSAGE,
-        u64(StreamId),
-        u16(Flags bor ?BOUNDARY),
-        u64(MessageId)
-    ]);
-message({EventTime, Message}, MessageId, StreamId, Flags) ->
-    frame([
-        ?MESSAGE,
-        u64(StreamId),
-        u16(Flags bor ?EVENT_TIME),
-        u64(MessageId),
-        u64(EventTime),
-        Message
-    ]);
-message(Message, MessageId, StreamId, Flags) ->
+    MessageId :: non_neg_integer(),
+    EventTime :: non_neg_integer(),
+    Key :: string(),
+    Message :: iodata().
+message(Flags, StreamId, MessageId, EventTime, Key, Message) ->
     frame([
-        ?MESSAGE,
-        u64(StreamId),
         u16(Flags),
-        u64(MessageId),
-        Message
+        u64(StreamId),
+        if Flags band ?EPHEMERAL /= 0 then
+            <<>>
+        else
+            u64(MessageId)
+        end,
+        if Flags band ?EVENT_TIME /= 0 then
+            u64(EventTime)
+        else
+            <<>>
+        end,
+        if Flags band ?KEY /= 0 then
+            short_bytes(Key)
+        else
+            <<>>
+        end,
+        if Flags band ?BOUNDARY /= 0 then
+            <<>>
+        else
+            Message
+        end
```

Comment:
Since this is a "spec" it might be worth writing a function for this. Maybe along the lines of:

```erlang
flags_set_or_nil(Flags, Bits, Value) when Flags band Bits /= 0  -> Value;
flags_set_or_nil(_, _, _) -> [].
```
(Alternate name ideas: masked/3, flagged/3)

That is if we keep the pseudo-code in Erlang for now (I had adopted it for convenience).

Reply:
With the differences in `<<>>`/`[]` going in the flag-set or flag-clear case, I caved in to cut-and-paste.